lib/documents.js

```diff
@@ -0,0 +1,259 @@
+const idEncoder = new IdEncoder({
+  encoding: 'base58',
+  fixedLength: true,
+  multibase: true
+});
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['tokenization-document']);
+
+  // `externalIdHash` is a shard key
+  await promisify(database.createIndexes)([{
+    // there may be multiple documents that match the same external ID hash
+    // and queries must be supported to find them all...
+    // but do not allow multiple `internalId`s for the same `hash` and
+    // `externalIdHash`
+    collection: 'tokenization-document',
+    fields: {'document.externalIdHash': 1, 'document.hash': 1},
+    options: {unique: true, background: false}
+  }, {
+    // automatically expire documents with an `expires` date field
+    collection: 'tokenization-document',
+    fields: {'document.expires': 1},
+    options: {
+      unique: false,
+      background: false,
+      expireAfterSeconds: 0
+    }
+  }]);
+});
+
+/**
+ * Registers a `document` associated with an entity that is unambiguously
+ * identified via `externalId`. Multiple documents associated with the same
+ * entity may be registered by using the same `externalId`; `externalId`
+ * must be determined by the application.
+ *
+ * An `internalId` will be generated to unambiguously refer to the same
+ * entity referenced by `externalId`. This `internalId` will be randomly
+ * generated to prevent unwanted correlation -- and the `externalId` that
+ * is given will be HMAC'd using key material that resides in an external
+ * system before storing it is stored. This approach ensures that a stolen
+ * database on its own will not reveal the correlation between a particular
+ * `externalId` and an `internalId`. This `internalId` will persist and be
+ * reused across all documents registered using the same `externalId`.
+ *
+ * The `document` will be encrypted to the given `recipients` before it is
+ * stored. It will also be marked for expiration using the given `ttl`
+ * parameter. Once all of the registered documents for a given `externalId`
+ * have expired, the `externalId` and its associated `internalId` will be
+ * removed such that if the same `externalId` is given in the future, a
+ * new `internalId` will be generated.
+ *
+ * Optional `registrationData` can be provided. This information *will* be
+ * stored in the clear in the database without any HMACing. Therefore, it
+ * should not include any information that would unambiguously identify
+ * the entity associated with `externalId`. This should be considered
+ * application-specific meta data about the document registration.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.externalId - An application-specific string that
+ *   unambiguously identifies a particular entity to the application.
+ * @param {object} options.document - The document to register.
+ * @param {array} options.recipients - A list of recipients identified by
```

Comment:
```suggestion
 * @param {object[]} options.recipients - A list of recipient JWK objects, identified by
```
(Just wanted to specify that it has to be `{ header: { kid: string, alg: string } }` objects, not strings.)

Reply:
I don't think these are JWK objects. They are containers of JWE headers that have JWKs in them.